Assertion failure in ImplicitSearch

```scala

class Foo{
  val default = this
  def foo(a: Int)(b: Foo = default): b.type = b
  val s: default.type = foo(1)
}
```

```
xception in thread "main" java.lang.AssertionError: assertion failed: found: this.foo(1): (b: Foo)Foo(b), expected: Foo(Foo.this.default)
	at scala.Predef$.assert(Predef.scala:165)
	at dotty.tools.dotc.typer.Implicits$ImplicitSearch.<init>(Implicits.scala:454)
	at dotty.tools.dotc.typer.Implicits$$anonfun$inferImplicit$1$$anonfun$apply$17.apply(Implicits.scala:429)
	at dotty.tools.dotc.typer.Implicits$$anonfun$inferImplicit$1$$anonfun$apply$17.apply(Implicits.scala:425)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:148)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:53)
	at dotty.tools.dotc.typer.Implicits$$anonfun$inferImplicit$1.apply(Implicits.scala:425)
	at dotty.tools.dotc.typer.Implicits$$anonfun$inferImplicit$1.apply(Implicits.scala:420)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:36)
	at dotty.tools.dotc.typer.Implicits$class.inferImplicit(Implicits.scala:420)
	at dotty.tools.dotc.typer.Typer.inferImplicit(Typer.scala:58)
	at dotty.tools.dotc.typer.Implicits$$anonfun$inferView$1.apply(Implicits.scala:403)
	at dotty.tools.dotc.typer.Implicits$$anonfun$inferView$1.apply(Implicits.scala:397)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:36)
	at dotty.tools.dotc.typer.Implicits$class.inferView(Implicits.scala:396)
	at dotty.tools.dotc.typer.Typer.inferView(Typer.scala:58)
	at dotty.tools.dotc.typer.Typer.adaptToSubType$1(Typer.scala:1401)
	at dotty.tools.dotc.typer.Typer.adaptNoArgs$1(Typer.scala:1361)
	at dotty.tools.dotc.typer.Typer.adaptInterpolated(Typer.scala:1440)
	at dotty.tools.dotc.typer.Typer$$anonfun$adapt$1$$anonfun$apply$15.apply(Typer.scala:1185)
	at dotty.tools.dotc.typer.Typer$$anonfun$adapt$1$$anonfun$apply$15.apply(Typer.scala:1182)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:148)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:53)
	at dotty.tools.dotc.typer.Typer$$anonfun$adapt$1.apply(Typer.scala:1182)
	at dotty.tools.dotc.typer.Typer$$anonfun$adapt$1.apply(Typer.scala:1182)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:36)
	at dotty.tools.dotc.typer.Typer.adapt(Typer.scala:1181)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1090)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1088)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:148)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:53)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1088)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1127)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedValDef$1.apply(Typer.scala:883)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedValDef$1.apply(Typer.scala:877)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:36)
	at dotty.tools.dotc.typer.Typer.typedValDef(Typer.scala:877)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1025)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1078)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1090)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1088)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:148)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:53)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1088)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1112)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1123)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedClassDef$1.apply(Typer.scala:924)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedClassDef$1.apply(Typer.scala:907)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:36)
	at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:907)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1031)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1078)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1090)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1088)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:148)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:53)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1088)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1112)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1123)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedPackageDef$1.apply(Typer.scala:971)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedPackageDef$1.apply(Typer.scala:962)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:36)
	at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:962)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1068)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1080)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1090)
	at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:1088)
	at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:148)
	at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:53)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1088)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1127)
	at dotty.tools.dotc.typer.FrontEnd$$anonfun$typeCheck$1.apply$mcV$sp(FrontEnd.scala:42)
	at dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:19)
	at dotty.tools.dotc.typer.FrontEnd.typeCheck(FrontEnd.scala:40)
	at dotty.tools.dotc.typer.FrontEnd$$anonfun$runOn$3.apply(FrontEnd.scala:53)
	at dotty.tools.dotc.typer.FrontEnd$$anonfun$runOn$3.apply(FrontEnd.scala:53)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at dotty.tools.dotc.typer.FrontEnd.runOn(FrontEnd.scala:53)
	at dotty.tools.dotc.Run$$anonfun$compileUnits$1$$anonfun$apply$mcV$sp$1.apply(Run.scala:59)
	at dotty.tools.dotc.Run$$anonfun$compileUnits$1$$anonfun$apply$mcV$sp$1.apply(Run.scala:56)
	at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
	at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply$mcV$sp(Run.scala:56)
	at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply(Run.scala:52)
	at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply(Run.scala:52)
	at dotty.tools.dotc.util.Stats$.monitorHeartBeat(Stats.scala:69)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:52)
	at dotty.tools.dotc.Run.compileSources(Run.scala:49)
	at dotty.tools.dotc.Run.compile(Run.scala:33)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:21)
	at dotty.tools.dotc.Driver.process(Driver.scala:44)
	at dotty.tools.dotc.Driver.main(Driver.scala:48)
	at dotty.tools.dotc.Main.main(Main.scala)
```

--- a/src/dotty/tools/dotc/typer/Typer.scala
+++ b/src/dotty/tools/dotc/typer/Typer.scala
@@ -1343,6 +1343,10 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
 
     def methodStr = err.refStr(methPart(tree).tpe)
 
+    def missingArgs = errorTree(tree,
+      d"""missing arguments for $methodStr
+         |follow this method with `_' if you want to treat it as a partially applied function""".stripMargin)
+
     def adaptOverloaded(ref: TermRef) = {
       val altDenots = ref.denot.alternatives
       typr.println(i"adapt overloaded $ref with alternatives ${altDenots map (_.info)}%, %")
@@ -1475,12 +1479,11 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
         else if (wtp.isImplicit)
           err.typeMismatch(tree, pt)
         else
-          errorTree(tree,
-            d"""missing arguments for $methodStr
-               |follow this method with `_' if you want to treat it as a partially applied function""".stripMargin)
+          missingArgs
       case _ =>
         if (tree.tpe <:< pt) tree
         else if (ctx.mode is Mode.Pattern) tree // no subtype check for pattern
+        else if (wtp.isInstanceOf[MethodType]) missingArgs
         else {
           typr.println(i"adapt to subtype ${tree.tpe} !<:< $pt")
           //typr.println(TypeComparer.explained(implicit ctx => tree.tpe <:< pt))

--- a/test/dotc/tests.scala
+++ b/test/dotc/tests.scala
@@ -157,6 +157,7 @@ class tests extends CompilerTest {
   @Test def neg_i941 = compileFile(negDir, "i941", xerrors = 3)
   @Test def neg_finalSealed = compileFile(negDir, "final-sealed", xerrors = 2)
   @Test def neg_i705 = compileFile(negDir, "i705-inner-value-class", xerrors = 7)
+  @Test def neg_i803 = compileFile(negDir, "i803", xerrors = 2)
   @Test def neg_i866 = compileFile(negDir, "i866", xerrors = 2)
   @Test def neg_i974 = compileFile(negDir, "i974", xerrors = 2)
   @Test def neg_moduleSubtyping = compileFile(negDir, "moduleSubtyping", xerrors = 4)

--- a/tests/neg/i803.scala
+++ b/tests/neg/i803.scala
@@ -1,0 +1,11 @@
+class Foo{
+  val default = this
+  def foo(a: Int)(b: Foo = default): b.type = b
+
+  def bar(b: Foo = default): b.type = b
+  val x: Foo = bar() // ok
+  val x2: Foo = foo(1)() // ok
+
+  val s: Foo = foo(1) // error
+  val s2: default.type = foo(1) // error
+}

